cgp/genome.py

```diff
@@ -354,77 +357,75 @@ def mutate(self, mutation_rate: float, rng: np.random.RandomState):
         active_regions = graph.determine_active_regions()
-
         dna = list(self._dna)
-
-        successful_mutations = 0
+        dna_array = np.array(dna)
         only_silent_mutations = True
-        while successful_mutations < n_mutations:
 
-            gene_idx = rng.randint(0, self._n_genes)
-            region_idx = gene_idx // self._length_per_region
+        selected_gene_indices = np.nonzero(np.random.rand(len(dna)) < mutation_rate)
 
-            if self._is_input_region(region_idx):
-                continue  # nothing to do here
+        for (gene_idx, gene) in zip(selected_gene_indices[0], dna_array[selected_gene_indices]):
 
-            elif self._is_output_region(region_idx):
-                success = self._mutate_output_region(dna, gene_idx, rng)
-                if success:
-                    silent = False
-                    only_silent_mutations = only_silent_mutations and silent
-                    successful_mutations += 1
+            silent = True
+            region_idx = gene_idx // self._length_per_region
+
+            permissible_values = self._determine_permissible_values(gene_idx, gene, region_idx)
+            if not len(permissible_values) == 0:
 
-            elif self._is_hidden_region(region_idx):
-                silent = self._mutate_hidden_region(dna, gene_idx, active_regions, rng)
-                only_silent_mutations = only_silent_mutations and silent
+                dna[gene_idx] = rng.choice(permissible_values)
+                silent = region_idx not in active_regions
 
-            else:
-                assert False  # should never be reached
+            only_silent_mutations = only_silent_mutations and silent
 
         self.dna = dna
         return only_silent_mutations
 
-    def _mutate_output_region(self, dna: List[int], gene_idx: int, rng: np.random.RandomState):
-        assert self._is_gene_in_output_region(gene_idx)
-
-        if not self._is_function_gene(gene_idx) and self._is_active_input_gene(gene_idx):
-            permissible_inputs = self._permissible_inputs_for_output_region()
-            dna[gene_idx] = rng.choice(permissible_inputs)
-            return True
-        else:
-            return False
+    def _determine_permissible_values(
+        self, gene_idx: int, gene: int, region_idx: int
+    ) -> List[int]:
 
-    def _mutate_hidden_region(
-        self, dna: List[int], gene_idx: int, active_regions: List[int], rng: np.random.RandomState
-    ) -> bool:
+        if self._is_input_region(region_idx):
+            return []  # input region have no other permissible values
 
-        assert self._is_gene_in_hidden_region(gene_idx)
+        elif self._is_hidden_region(region_idx):
+            return self._determine_permissible_values_hidden(gene_idx, gene, region_idx)
 
-        region_idx = gene_idx // self._length_per_region
-        silent_mutation = region_idx not in active_regions
+        else:
+            assert self._is_output_region(region_idx)
```

Comment:
why not another `elif` here, i.e. `elif self._is_output_region(region_idx)`?

Reply:
Just because it is shorter than `elif self._is_output_region(region_idx)` and `else False`

Reply:
I would prefer having an `elif` and `else` statement here, because this is more readable and maintainable. Same below.